### PR TITLE
[AVR] Replace broken 'avr-unknown-unknown' target with 'avr-unknown-gnu-atmega328' target

### PR DESCRIPTION
**Pull request by dylanmckay**
The `avr-unknown-unknown` target has never worked correctly, always trying to invoke
the host linker and failing. It aimed to be a mirror of AVR-GCC's
default handling of the `avr-unknown-unknown' triple (assume bare
minimum chip features, silently skip linking runtime libraries, etc).
This behaviour is broken-by-default as it will cause a miscompiled executable
when flashed.

This patch improves the AVR builtin target specifications to instead
expose only a 'avr-unknown-gnu-atmega328' target. This target system is
`gnu`, as it uses the AVR-GCC frontend along with avr-binutils. The
target triple ABI is 'atmega328'.

In the future, it should be possible to replace the dependency on
AVR-GCC and binutils by using the in-progress AVR LLD and compiler-rt support.
Perhaps at that point it would make sense to add an
'avr-unknown-unknown-atmega328' target as a better default when
implemented.

There is no current intention to add in-tree AVR target specifications for other
AVR microcontrollers - this one can serve as a reference implementation
for other devices via `rustc --print target-spec-json
avr-unknown-gnu-atmega328p`.

There should be no users of the existing 'avr-unknown-unknown' Rust
target as a custom target specification JSON has always been
recommended, and the avr-unknown-unknown target could never pass the
linking step anyway.

### CODE REVIEW

File: src/librustc_target/spec/avr_gnu_base.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        target_env: "".to_string(),
+        target_vendor: "unknown".to_string(),
+        target_c_int_width: 16.to_string(),
+        options: TargetOptions {
+            cpu: target_cpu.clone(),
+            exe_suffix: ".elf".to_string(),
+            linker: Some("avr-gcc".to_owned()),
+            pre_link_args: vec![(
+                LinkerFlavor::Gcc,
+                vec!["-Os".to_owned(), format!("-mmcu={}", target_cpu)],
```
Comment (petrochenkov) [inline src/librustc_target/spec/avr_gnu_base.rs:24]: Does `-Os` have any effect when `gcc` is used as a linker only? What `ld` option does it translate to?

File: src/librustc_target/spec/avr_gnu_base.rs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+            )]
+            .into_iter()
+            .collect(),
+            late_link_args: vec![(LinkerFlavor::Gcc, vec!["-lc".to_owned(), "-lgcc".to_owned()])]
+                .into_iter()
+                .collect(),
+            ..super::freestanding_base::opts()
```
Comment (petrochenkov) [inline src/librustc_target/spec/avr_gnu_base.rs:31]: Could you inline this and remove `freestanding_base.rs`?
It looks misleadingly generic, but is only used for AVR.

File: src/librustc_target/spec/avr_gnu_base.rs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            cpu: target_cpu.clone(),
+            exe_suffix: ".elf".to_string(),
+            linker: Some("avr-gcc".to_owned()),
+            pre_link_args: vec![(
+                LinkerFlavor::Gcc,
+                vec!["-Os".to_owned(), format!("-mmcu={}", target_cpu)],
+            )]
+            .into_iter()
+            .collect(),
+            late_link_args: vec![(LinkerFlavor::Gcc, vec!["-lc".to_owned(), "-lgcc".to_owned()])]
```
Comment (petrochenkov) [inline src/librustc_target/spec/avr_gnu_base.rs:28]: Linking libc is supposed to be done through the `libc` crate ideally.

**Comment by oli-obk** [comment]
cc @shepmaster @japaric 

cc @rust-lang/compiler this PR replaces a Tier 3 target that apparently never worked anyway with a controller specific tier 3 target that mostly serves as a reference target for others to copy and reuse

File: src/librustc_target/spec/avr_gnu_base.rs (hunk 1 above)
Comment (mati865) [reply src/librustc_target/spec/avr_gnu_base.rs:24]: At least when it comes to x86 linkers use `-O0` to `-O2` to reduce size, with `-O1` being default.

File: src/librustc_target/spec/avr_gnu_base.rs (hunk 3 above)
Comment (dylanmckay) [reply src/librustc_target/spec/avr_gnu_base.rs]: Upon further inspection, I don't believe libc is actually required. I copied the options over from the blink example, but it is totally stripped out when linking as there are no references to it in emitted Rust or LLVM code.

I've dropped the dependency in c4d3e03eb82719f86d20ae105a19c81ce80f5587

File: src/librustc_target/spec/avr_gnu_base.rs (hunk 2 above)
Comment (dylanmckay) [reply src/librustc_target/spec/avr_gnu_base.rs]: Agree, fixed in ffe98aa186312f9e8a74765f75e0f32757cc425a

File: src/librustc_target/spec/avr_gnu_base.rs (hunk 1 above)
Comment (dylanmckay) [reply src/librustc_target/spec/avr_gnu_base.rs]: I ended up removing it in c4d3e03eb82719f86d20ae105a19c81ce80f5587. Most AVR projects will still need to use a custom JSON target specification file, in which case, it is trivial to add custom link arguments that can be tailored to the project.

**Comment by nikomatsakis** [comment]
Sounds good to me, I say land it.

**Comment by oli-obk** [comment]
@dylanmckay can you rebase and fix the conflicts? Then we can move to merging this

**Comment by dylanmckay** [comment]
The branch has now been rebased.

**Comment by dylanmckay** [comment]
Strange, seemingly unrelated "command not found" error from bors https://github.com/rust-lang-ci/rust/runs/1032158434#step:23:22604 (`[run-make] run-make-fulldeps\pgo-indirect-call-promotion`)

**Comment by mati865** [comment]
It's spurious failure.

**Comment by sunfishcode** [comment]
Since `-gnu` matches an existing [environment type](https://llvm.org/doxygen/classllvm_1_1Triple.html#a83e907e55fa50e093caa96a0aff96201), does that imply that the `-atmega328` in the triple here refers to a new [object format type](https://llvm.org/doxygen/classllvm_1_1Triple.html#a83e907e55fa50e093caa96a0aff96201)?

**Comment by dylanmckay** [comment]
> Since -gnu matches an existing environment type, does that imply that the -atmega328 in the triple here refers to a new object format type?

It does not - here it is only meant to indicate that this triple specificaly targets the GNU toolchain/linker/libraries.

**Pull request merged**